core/src/program.rs

```diff
@@ -18,11 +18,112 @@
 
 //! Module for programs.
 
+use crate::{
+    ids::{MessageId, ProgramId},
+    message::DispatchKind,
+    pages::{numerated::tree::IntervalsTree, GearPage, WasmPage, WasmPagesAmount},
+    reservation::GasReservationMap,
+};
+use alloc::collections::BTreeSet;
+use primitive_types::H256;
 use scale_info::{
     scale::{Decode, Encode},
     TypeInfo,
 };
 
+/// Program in different states in storage.
+#[derive(Clone, Debug, Decode, Encode, PartialEq, Eq, TypeInfo)]
+pub enum Program<BlockNumber: Copy> {
+    /// Program in active state.
+    Active(ActiveProgram<BlockNumber>),
+    /// Program has been exited (gr_exit was called)
+    Exited(ProgramId),
+    /// Program has been terminated (`init` was failed)
+    Terminated(ProgramId),
+}
+
+impl<BlockNumber: Copy> Program<BlockNumber> {
+    /// Returns whether the program is active.
+    pub fn is_active(&self) -> bool {
+        matches!(self, Program::Active(_))
+    }
+
+    /// Returns whether the program is exited.
+    pub fn is_exited(&self) -> bool {
+        matches!(self, Program::Exited(_))
+    }
+
+    /// Returns whether the program is terminated.
+    pub fn is_terminated(&self) -> bool {
+        matches!(self, Program::Terminated(_))
+    }
+
+    /// Returns whether the program is active and initialized.
+    pub fn is_initialized(&self) -> bool {
+        matches!(
+            self,
+            Program::Active(ActiveProgram {
+                state: ProgramState::Initialized,
+                ..
+            })
+        )
+    }
+}
+
+/// Program is not an active one.
+#[derive(Clone, Debug, derive_more::Display)]
+#[display(fmt = "Program is not an active one")]
+pub struct InactiveProgramError;
+
+impl<BlockNumber: Copy> core::convert::TryFrom<Program<BlockNumber>>
+    for ActiveProgram<BlockNumber>
+{
+    type Error = InactiveProgramError;
+
+    fn try_from(prog_with_status: Program<BlockNumber>) -> Result<Self, Self::Error> {
+        match prog_with_status {
+            Program::Active(p) => Ok(p),
+            _ => Err(InactiveProgramError),
+        }
+    }
+}
+
+/// Active program in storage.
+#[derive(Clone, Debug, Decode, Encode, PartialEq, Eq, TypeInfo)]
+pub struct ActiveProgram<BlockNumber: Copy> {
```

Comment:
what about these?
```rust
#[codec(crate = codec)]
#[scale_info(crate = scale_info)]
```

Reply:
I'd used `gear-core` style here. Do you this there is any differences? `scale_info` used by default as `scale_info crate`, and in core we use naming `parity-scale-codec` instead of `codec`.

If you think this is must be refactored in core - let's create separate issue for that ?

Reply:
discussed